vnext/ReactUWP/Modules/AsyncStorageModuleWin32.cpp

```diff
@@ -0,0 +1,191 @@
+          // in turn allows JS user code to test for support of merge methods.
+          Method("clear", this, &AsyncStorageModuleWin32::clear),
+          Method("getAllKeys", this, &AsyncStorageModuleWin32::getAllKeys)};
+}
+
+void AsyncStorageModuleWin32::multiGet(folly::dynamic params, Callback callback) {
+  folly::dynamic result = folly::dynamic::array;
+  auto &&args = params[0];
+  if (args.size() != 0) {
+    Sqlite3Transaction transaction(m_db);
+    auto pStmt = PrepareStatement(u8"SELECT key, value FROM AsyncLocalStorage WHERE key = ?");
+    for (auto &&arg : args) {
+      BindString(pStmt, 1, arg.getString());
+      checkSQLiteResult(sqlite3_step(pStmt.get()));
+      auto key = reinterpret_cast<const char *>(sqlite3_column_text(pStmt.get(), 0));
+      if (!key)
+        checkSQLiteResult(sqlite3_errcode(m_db));
+      auto value = reinterpret_cast<const char *>(sqlite3_column_text(pStmt.get(), 1));
+      if (!value)
+        checkSQLiteResult(sqlite3_errcode(m_db));
+      result.push_back(folly::dynamic::array(std::string(key), std::string(value)));
+      checkSQLiteResult(sqlite3_reset(pStmt.get()));
+    }
+  }
+  callback({folly::dynamic(), result});
+}
+
+void AsyncStorageModuleWin32::multiSet(folly::dynamic params, Callback callback) {
+  auto &&args = params[0];
+  if (args.size() != 0) {
+    Sqlite3Transaction transaction(m_db);
+    auto pStmt = PrepareStatement(u8"INSERT OR REPLACE INTO AsyncLocalStorage VALUES(?, ?)");
+    for (auto &&arg : args) {
+      BindString(pStmt, 1, arg[0].getString());
+      BindString(pStmt, 2, arg[1].getString());
+      checkSQLiteResult(sqlite3_step(pStmt.get()));
+      checkSQLiteResult(sqlite3_reset(pStmt.get()));
+    }
+  }
+  callback({});
+}
+
+void AsyncStorageModuleWin32::multiRemove(folly::dynamic params, Callback callback) {
+  auto &&args = params[0];
+  if (args.size() != 0) {
+    Sqlite3Transaction transaction(m_db);
+    auto pStmt = PrepareStatement(u8"DELETE FROM AsyncLocalStorage WHERE key = ?");
+    for (auto &&arg : args) {
+      checkSQLiteResult(sqlite3_bind_text(pStmt.get(), 1, arg.getString().c_str(), -1, SQLITE_TRANSIENT));
+      checkSQLiteResult(sqlite3_step(pStmt.get()));
+      checkSQLiteResult(sqlite3_reset(pStmt.get()));
+    }
+  }
+  callback({});
+}
+
+void AsyncStorageModuleWin32::clear(folly::dynamic, Callback callback) {
+  Exec(u8"DELETE FROM AsyncLocalStorage");
+  callback({});
+}
+
+void AsyncStorageModuleWin32::getAllKeys(folly::dynamic, Callback callback) {
+  folly::dynamic jsRetVal = folly::dynamic::array;
+  auto getAllKeysCallback = [&](int cCol, char **rgszColText, char **) {
+    if (cCol >= 1) {
+      jsRetVal.push_back(rgszColText[0]);
+    }
+    return SQLITE_OK;
+  };
+
+  Exec(u8"SELECT key FROM AsyncLocalStorage", getAllKeysCallback);
+
+  if (jsRetVal.size() == 0) {
+    callback({"AsyncStorageError - No Keys Found", {}});
+  }
+}
+
+auto AsyncStorageModuleWin32::PrepareStatement(const char *stmt) -> Statement {
```

Comment:
Nit: why the trailing return here?

Reply:
Because otherwise I have to type `AsyncStorageModuleWin32::Statement` as the return type because C++ is a really bad language. I can do that if you prefer

Reply:
Would personally prefer doing without the trailing return. seeing `auto` as a return type makes me initially guess we're doing something funky with generics or using an expressible type like a lambda.